blog/modules/ROOT/pages/using-types.adoc

```diff
@@ -0,0 +1,220 @@
+= Taking types to the next level
+
+:use-link-attrs:
+
+// tag::byline[]
+++++
+<div class="blog-byline">
+++++
+by link:https://github.com/stackoverflow[Islon Scherer] on October 1st, 2024
+++++
+</div>
+++++
+// end::byline[]
+
+// tag::excerpt[]
+One of the main points of using Pkl is to describe what you configuration looks like.
+But an often overlooked design goal (arguably, the most important), is to
+*forbid* invalid configurations. Pkl types allow a great degree of freedom
+and power in constraining what you data looks like.
+// end::excerpt[]
+
+Some developers, especially those coming from more traditional languages like
+Go or Java, may have a pre-defined notion of how types can help them.
+So let's have a look how Pkl can go beyond most traditional types.
+
+== Beyond Strings
+
+`String` is a type we tend to see everywhere, this is especially true in configuration
+languages, where a `String` can be used as a "joker" type for many different values.
+
+=== Enumerating possibilities
+
+A typical usage of Strings is to define a finite set of values that you property can take:
+
+[source,pkl]
+----
+/// The environment where this service is running.
+///
+/// Valid possibilities are "dev", "qa" and "prod".
+env: String
+----
+
+This works, but has the problem that users of your template can provide invalid data, like `"foo"`.
+So let's do better:
+
+[source,pkl]
+----
+/// The environment where this service is running.
+env: "dev"|"qa"|"prod"
```

Comment:
Not sure whether _here_ is the place, but one pattern I've found very useful is to have a `|String` at then end of such a union type; for correctness - and in interpreter terms, generally - it basically means `env: String`, but it informs autocomplete.

Reply:
Yeah, that's an useful pattern, but the point of this example is exactly to have a closed list of possibilities. Maybe I can add it in the "more string goodies" section.